src/main/java/org/galatea/starter/service/IexService.java

```diff
@@ -50,28 +60,177 @@
   /**
    * Get the historical price for each symbol passed in for the date passed in.
    *
-   * @param symbol list of symbols to get historical price for.
+   * @param symbol Stock symbol to get historical price for.
    * @param range the range of time  (ex. "5m", "ytd" ) to get previous data (Optional).
    * @param date the date from which we would want to get the previous data from (Optional).
-   * Note: If neither optional parameter is used, the system will default to range = 1m.
-   * - default behavior of Iex Cloud API
-   * Note: If both optional parameters are provided, the system will default to the date.
-   * - can be changed later if needed: IexCloud will not allow both options in call to client
-   * @return a IexHistoricalPrices objects for the given symbols.
+   *      Note: If neither optional parameter is used, the system will default to range = 1m.
+   *      - default behavior of Iex Cloud API
+   *      Note: If both optional parameters are provided, the system will default to the date.
+   *      - can be changed later if needed: IexCloud will not allow both options in call to client
+   * @return a list of IexHistoricalPrices objects for the given symbols.
    */
   public List<IexHistoricalPrices> getHistoricalPricesForSymbol(
-      final String symbol, String range, String date) {
-    if (symbol==null) {
+      final String symbol, final String range, final String date) {
+
+    if (symbol == null) {
       throw new IllegalArgumentException("No Stock Symbol Provided.");
-    } else if (range==null && date==null) {
-      String rangeVal = "1m";
-      return iexClient.getHistoricalPricesForSymbolByRange(symbol, rangeVal);
-    } else if (date==null) {
-      return iexClient.getHistoricalPricesForSymbolByRange(symbol, range);
+    } else if (!isSymbolInDatabase(symbol)) {
+      //if symbol isn't in database, call Iex normally
+      return getHistoricalPricesFromIex(symbol, range, date);
+    }
+    else if (date != null) {
+      // Only returns a list of one object
+      return getHistoricalPriceBySymbolAndDate(symbol, date);
     } else {
-      return iexClient.getHistoricalPricesForSymbolByDate(symbol, date);
+      // Parse range into a list of dates and run them through getHistoricalPriceBySymbolAndDate
+      List<String> datesToLookThrough = rangeToDateList(range);
+      List<IexHistoricalPrices> outList = new ArrayList<>();
+      for (String dateVal : datesToLookThrough) {
+        List<IexHistoricalPrices> out = getHistoricalPriceBySymbolAndDate(symbol, dateVal);
```

Comment:
The way this is coded, it would result in many calls to IEX for each date. This works, but it seems a lot less efficient than querying for all missing dates in one query. Does IEX have an API for that, or would you need to build a range query based on the dates?

Reply:
I know it's inefficient, I just can't find an easy way around it. For the historical prices endpoint, the date or range input is a Path Variable, not a parameter, so we can only pass in once per call. So, based on the documentation for this endpoint, the range parameter is the only way to do a batch call. The range parameter is also a trailing range parameter, so I could not make a call for something like "find previous 5 months of data ending on 2021-12-31".

Reply:
I could try to use another endpoint (possibly the time series endpoint) to do this instead, but I'm not sure if we want to add that extra level of complexity to the system. I'm happy to try it if you would like me to though.